sockeye/encoder.py

```diff
@@ -413,3 +470,173 @@ def get_rnn_cells(self) -> List[mx.rnn.BaseRNNCell]:
         Returns a list of RNNCells used by this encoder.
         """
         return self.forward_rnn.get_rnn_cells() + self.reverse_rnn.get_rnn_cells()
+
+
+class CharacterSequenceEncoder(Encoder):
+    """
+    An encoder that maps a sequence of character embeddings to a shorter
+    sequence of segment embeddings using convolutional, pooling, and highway
+    layers.
+        * "Fully Character-Level Neural Machine Translation without Explicit Segmentation"
+          Jason Lee; Kyunghyun Cho; Thomas Hofmann (https://arxiv.org/pdf/1610.03017.pdf)
+
+    :param num_embed: Input embedding size.
+    :param max_filter_width: Maximum filter width for convolutions.
+    :param num_filters: Number of filters of each width.
+    :param pool_stride: Stride for pooling layer after convolutions.
+    :param num_highway_layers: Number of highway layers for segment embeddings.
+    :param prefix: Name prefix for symbols of this encoder.
+    :param dropout: Dropout probability.
+    """
+
+    def __init__(self,
+                 num_embed: int,
+                 max_filter_width: int = 8,
+                 num_filters: List[int] = None,
+                 pool_stride: int = 5,
+                 num_highway_layers: int = 4,
+                 prefix: str = C.CHAR_SEQ_ENCODER_PREFIX,
+                 dropout: float = 0.):
+        if not num_filters:
+            num_filters = [200, 200, 250, 250, 300, 300, 300, 300]
+        assert len(num_filters) == max_filter_width, "num_filters must have max_filter_width elements."
```

Comment:
please change to check_condition to make David happy for not using assertions to validate user input :)